core/src/main/java/org/apache/iceberg/PartitionEntry.java

```diff
@@ -0,0 +1,361 @@
+            9, Column.EQUALITY_DELETE_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.optional(10, Column.TOTAL_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(11, Column.LAST_UPDATED_AT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            12, Column.LAST_UPDATED_SNAPSHOT_ID.name(), Types.LongType.get()));
+  }
+
+  private static Schema prepareAvroSchema(Types.StructType partitionType) {
+    return AvroSchemaUtil.convert(icebergSchema(partitionType), "partitionEntry");
+  }
+
+  public static class Builder {
+    private PartitionData partitionData;
+    private int specId;
+    private long dataRecordCount;
+    private int dataFileCount;
+    private long dataFileSizeInBytes;
+    private long posDeleteRecordCount;
+    private int posDeleteFileCount;
+    private long eqDeleteRecordCount;
+    private int eqDeleteFileCount;
+    private long totalRecordCount;
+    private long lastUpdatedAt;
+    private long lastUpdatedSnapshotId;
+
+    private Builder() {}
+
+    public Builder withPartitionData(PartitionData newPartitionData) {
+      this.partitionData = newPartitionData;
+      return this;
+    }
+
+    public Builder withSpecId(int newSpecId) {
+      this.specId = newSpecId;
+      return this;
+    }
+
+    public Builder withDataRecordCount(long newDataRecordCount) {
+      this.dataRecordCount = newDataRecordCount;
+      return this;
+    }
+
+    public Builder withDataFileCount(int newDataFileCount) {
+      this.dataFileCount = newDataFileCount;
+      return this;
+    }
+
+    public Builder withDataFileSizeInBytes(long newDataFileSizeInBytes) {
+      this.dataFileSizeInBytes = newDataFileSizeInBytes;
+      return this;
+    }
+
+    public Builder withPosDeleteRecordCount(Long newPosDeleteRecordCount) {
+      this.posDeleteRecordCount = newPosDeleteRecordCount;
+      return this;
+    }
+
+    public Builder withPosDeleteFileCount(Integer newPosDeleteFileCount) {
+      this.posDeleteFileCount = newPosDeleteFileCount;
+      return this;
+    }
+
+    public Builder withEqDeleteRecordCount(Long newEqDeleteRecordCount) {
+      this.eqDeleteRecordCount = newEqDeleteRecordCount;
+      return this;
+    }
+
+    public Builder withEqDeleteFileCount(Integer newEqDeleteFileCount) {
+      this.eqDeleteFileCount = newEqDeleteFileCount;
+      return this;
+    }
+
+    public Builder withTotalRecordCount(Long newTotalRecordCount) {
+      this.totalRecordCount = newTotalRecordCount;
+      return this;
+    }
+
+    public Builder withLastUpdatedAt(Long newLastUpdatedAt) {
+      this.lastUpdatedAt = newLastUpdatedAt;
+      return this;
+    }
+
+    public Builder withLastUpdatedSnapshotId(Long newLastUpdatedSnapshotId) {
+      this.lastUpdatedSnapshotId = newLastUpdatedSnapshotId;
+      return this;
+    }
+
+    public PartitionEntry newInstance() {
+      return new PartitionEntry();
+    }
+
+    public PartitionEntry build() {
```

Comment:
using the builder instead of `Immutables` because these objects will be mutable in a partition stats map. Will be adding the update functions during impl.